### DISCUSSION TITLE
Ошибка 403 при использовании Source.getArtistsTracks

### DISCUSSION THREAD

**Discussion by Bondian** [Chimildic/goofy#106] [Помощь с ошибкой] [answered]
В ранее работавшей функции getArtistsTracks стали появляться ошибки. Подумал, что появилось обновление и работа функций изменилась, и точно. Обновился до 1.4.9. Но ошибка не исчезла. Зашёл в [эту](https://github.com/Chimildic/goofy/discussions/104) свежую тему. Скопировал там код и тоже ошибка...
14:29:44	Примечание	Выполнение начато
14:29:51	Ошибка	URL: https://api.spotify.com/v1/albums/12IkYAeG3Oj6zICP2NI75A/tracks 
Code: 403 
Content: Forbidden
14:29:47	Ошибка	
TypeError: Cannot read property 'items' of undefined
getItemsByPath	@ library.gs:2952
getAlbumTracks	@ library.gs:324
(функция без названия)	@ library.gs:270
getArtistsTracks	@ library.gs:269
createPlaylistForArtist	@ Без названия.gs:2

Права доступа обновлены...

**Comment by koojichan** [comment]
А ты URL правильно ввёл?

**Reply by Bondian** [reply]
Я не вводил URL. Скопировал-вставил...

**Reply by koojichan** [reply]
> Я не вводил URL. Скопировал-вставил...

Конкретно в коде из свежей темы, не надо вводить ссылку. Чтобы получить id, нужно зайти на страницу плейлиста, артиста, или пользователя в Spotify и нажать поделиться. Так получим ссылку на плейлист. Из нее нужно достать id. Если открыть меню поделиться и нажать кнопку Alt, Spotify предложит скопировать URI, а не ссылку.

> let mixTracks = Source.getTracks([
    { name: 'Микс дня 1', id: '491ZfFnGxaBF445JOhhxiO' },
    { name: 'Микс дня 2', id: '426ZfFnGxaBF445JfOJefE' },
]);

// Ошибки!
// { name: 'Микс дня 1', id: 'https://open.spotify.com/playlist/491ZfFnGxaBF445JOhhxiO?si=343F7972b107494a' },
// { name: 'Микс дня 1', id: 'spotify:playlist:426ZfFnGxaBF445JfOJefE' },

**Reply by Bondian** [reply]
В данном коде уже содержится id. Для тестирования работоспособности кода меня устраивает исполнитель nothing,nowhere. Зачем его менять?

**Reply by koojichan** [reply]
> В данном коде уже содержится id. Для тестирования работоспособности кода меня устраивает исполнитель nothing,nowhere. Зачем его менять?

Попробуй тогда код который я использовал, у меня он работает прекрасно.

```jsx
function createPlaylistForArtist() {
  let tracks = Source.getArtistsTracks({
    artist: {
      followed_include: false,
      include: [
        { id: '20UYCAvAHJ1WqrCElptD7O' },
      ]
    },
    album: {
      groups: 'album,single',
    },
  });

  Playlist.saveWithReplace({
    // id: 'id',
    name: 'All of Kotori ^~^',
    tracks: tracks,
  })
};
```

**Reply by Bondian** [reply]
Если Вы говорите об URL, указанном в сообщении, то это скопированный лог выполнения того кода...

**Reply by koojichan** [reply]
> Если Вы говорите об URL, указанном в сообщении, то это скопированный лог выполнения того кода...

Понял, тут уже мои полномочия всё.

**Reply by Bondian** [reply]
Скопировал Ваш код. Снова ошибка

15:10:53	Примечание	Выполнение начато
15:10:59	Ошибка	URL: https://api.spotify.com/v1/albums/3wEq76CrSEAMpNZw2D5SeI/tracks 
Code: 403 
Content: Forbidden
15:10:54	Ошибка	
TypeError: Cannot read property 'items' of undefined
getItemsByPath	@ library.gs:2952
getAlbumTracks	@ library.gs:324
(функция без названия)	@ library.gs:270
getArtistsTracks	@ library.gs:269
createPlaylistForArtist	@ Без названия.gs:2

**Reply by Bondian** [reply]
После нескольких попыток выполнение прошло без ошибок ))) Повторные выполнения снова выдают ошибки...

**Reply by koojichan** [reply]
Тоже сейчас попробовал, выдало одну ошибку, затем ещё одна попытка и все успешно.

**Comment by Chimildic** [comment, marked as answer]
> 403 Forbidden

Ошибка о запрете доступа. На [форуме Spotify](https://community.spotify.com/t5/Spotify-for-Developers/Some-API-endpoints-randomly-return-error-403/td-p/5233575) есть тема с воскресенья о ее случайном появлении в разных вызовах не смотря на ключи и доступ. 

> TypeError: Cannot read property 'items' of undefined

Ошибка в следствии отсутствия контента из-за предыдущей. 

Можно предусмотреть пару костылей, но это не решит проблему. Текущий код способен работать без изменений, когда Spotify поправит ответы на запросы.

**Reply by Bondian** [reply]
Спасибо за разъяснение
